src/cloud-search-index-storage.js

```diff
@@ -0,0 +1,210 @@
+
+export class CloudSearchIndexStorage {
+  #FIELDS_TO_INDEX = new Set(['objectID', 'assetIdentity', 'contentHash', 'sourceName', 'sourceType', 'thumbnailHash', 'file', 'companyId', 'spaceId', 'type', 'tags', 'ocrTags', 'caption', 'width', 'height', 'color', 'sourceId', 'sourceUrl', 'sourceAssetId', 'sourcePath']);
+  #INDEX_PREFIX = 'company';
+
+  constructor(context, companyId) {
+    // TODO: Job document should specify the companyId which should be used as the index name
+    this.indexName = context.env.ALGOLIA_CI_INDEX || `${this.#INDEX_PREFIX}-${companyId}`;
+    console.log('Using Search Index', this.indexName);
+    this.index = algoliasearch(
+      context.env.ALGOLIA_APP_NAME,
+      context.env.ALGOLIA_API_KEY,
+    ).initIndex(this.indexName);
+  }
+
+  withLog(logger) {
+    this.log = logger;
+    return this;
+  }
+
+  /**
+   * Strong Existence Check
+   *
+   * Needs to have both the sourceId and the contentHash (source binary hash) matching
+   * @param {Object} query
+   * @returns Object of hits if there is a hit or undefined if it doesn't exist
+   */
+  async exists(query) {
+    const facetFilters = [];
+    // if no contentHash, we don't want to match just on the sourceId
+    // becuase that is not reliable so we will not return a match
+    if (!query.contentHash) {
+      console.log('Missing contentHash, no existence in the record storage.');
+      return undefined;
+    }
+
+    facetFilters.push(`contentHash:${query.contentHash}`);
+
+    if (query?.sourceAssetId) {
+      facetFilters.push(`sourceAssetId:${query.sourceAssetId}`);
+    }
+
+    const searchResult = await this.index.search(
+      '',
+      {
+        facetFilters,
+      },
+    );
+    if (searchResult?.nbHits > 0) {
+      return searchResult?.hits;
+    }
+    return false;
+  }
+
+  /**
+   * Retrieve a single index record via the record's identifier (objectID).
+   * @param {*} objectID - the unique identifier for the index record
+   * @returns the raw result of the query; will contain the record if it exists.
+   */
+  async get(objectID) {
+    return this.index.search(
+      '',
+      {
+        facetFilters: [
+          `objectID:${objectID}`,
+        ],
+      },
+    );
+  }
+
+  /**
+   * Save the information in the provided object to the search index as an index record.
+   *
+   * Not all fields provided in the object will be saved to the index. This method takes the values
+   * from the provided object that are specified to be used in the index and saves a record with
+   * those values; other values are ignored.
+   * @param {Object} doc the object to store in the index.
+   * @returns An object containing the saved record's ObjectID.
+   */
+  async save(doc) {
+    const indexRecord = Object.keys(doc).reduce((obj, key) => {
+      const result = clone(obj);
+      if (this.#FIELDS_TO_INDEX.has(key)) {
+        result[key] = doc[key];
+      }
+      return result;
+    }, {});
+    if (!indexRecord?.objectID) {
```

Comment:
This was moved out of this class in order to make sure we use the same id in the index record and in the metadata record.  See https://github.com/adobe/content-lake-ingestor-service/blob/036775b4fd7ff86e1e2848e9479a2f1d3789fc49/src/storage/cloud-search-index-storage.js#L96 and https://github.com/adobe/content-lake-ingestor-service/blob/036775b4fd7ff86e1e2848e9479a2f1d3789fc49/src/ingestor.js#L315

Reply:
Ok, so in our usage with the ingestor, we should not ever hit this case..
however, as a standalone algolia client, i think we should keep it. wdyt?

Reply:
Would this be worth adding a log statement since it _shouldn't_ happen?